docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+
+## Rationale
+Users can choose to store expiry time explicitly in the table schema and run delete on such tables periodically to reclaim space. This approach not only post burden on users but also introduced unnecessary load on TiDB to execute such updates and run GC at background to eventually reclaim space. The extra load is especially high when there is lots of data expiring within a short period of time.
+Another way to quickly remove expired data is managing all data in a carefully designed partition table. Bucketing data according to their update time makes it possible to quickly delete all expired records within a partition with a simple `TRUNCATE`. Compared to TTL Table's approach, these two options are error prone and suboptimal.
+
+## Compatibility and Migration Plan
+Due to the fact that TiDB by itself prohibits either converting ordinary table to partition table or between different partition table types, it is only possible to convert existing table to TTL table with `ROW` granularity. A TTL table with `PARTITION` granularity is implemented as a special partition table therefore conflicts with any other partition table type. Newly created tables can choose either `ROW` or `PARTITION` as a trade off between accuracy of reclaiming time and efficiency of garbage collecting expired data.
+
+## Implementation
+
+### `ROW`
+By associating TTL configuration with the key range of a TTL table in PD and distributing such configuration to all TiKV instances within the cluster, TiKV can utilize TTL settings during GC process to collect data that are expired. To avoid inconsistency caused by reclaiming progress difference between index and actual record data. The lifetime of record range is one GC interval longer than it is for index range, therefore TiDB will not see missing record data during table lookup.
+
+### `PARTITION`
+TiDB DDL master maintains a new periodic task to rollover partitions of a TTL table by allocating a new partition as current writing partition and truncate the oldest partition that has already passed its lifetime. Truncating partition is a `O1` operation that has nothing to do with the actual number of records within such partition. Therefore can be archived without blocking DDL for a noticeable amount of time with negligible amount of work at the background.
+
+## Testing Plan
+TBD
```

Comment:
could you elaborate more about:

1. compatibility tests
	- compatibility with other features, like MVCC GC, partition table
	- compatibility with other internal components, like Parser, DDL, Privilege, Statistics
	- compatibility with other external components, like PD, TiKV, TiFlash, BR, TiCDC, Dumpling
	- upgrade compatibility
	- downgrade compatibility
2. functional tests
	- to ensure the basic feature function works as expected?
3. scenario tests
	- to ensure this feature works as expected in some common scenarios
4. benchmark tests
	- to measure the timeliness of TTL mechanism
	- to measure the influence on the online workload when TTL triggered

Reply:
Sure, Iet me add more about these.